### PR TITLE
[Feature] Implement Move Multiple Items

### PR DESCRIPTION
**Pull request by catapultam-habeo**
# Description

This implements ctrl-click movement of bag contents for ROF+ clients.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# Testing

There are three behaviors that are introduced here;
* CTRL + left click -> drops an item into a bag without opening it.
* CTRL + left click -> adds the contents of a bag (on cursor) into another bag without opening it
* CTRL + right click -> swaps the contents of two bags if a bag is on your cursor and you ctrl-right click on another bag.

I don't know how to attach images to demonstrate this.

I would furthermore like some assistance testing this.

Clients tested: 
* ROF2

# Checklist

- [x] I have tested my changes
- [X] I have performed a self-review of my code. Ensuring variables, functions and methods are named in a human-readable way, comments are added only where naming of variables, functions and methods can't give enough context.
- [ ] I have made corresponding changes to the documentation (if applicable, if not delete this line)
- [X] I own the changes of my code and take responsibility for the potential issues that occur

### CODE REVIEW

File: zone/client_packet.cpp [hunk 1]
```diff
@@ -10859,7 +10859,50 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
 void Client::Handle_OP_MoveMultipleItems(const EQApplicationPacket *app)
 {
-	Kick("Unimplemented move multiple items"); // TODO: lets not desync though
+	if (m_ClientVersionBit & EQ::versions::maskRoF2AndLater) {		
+		if (!CharacterID()) {
+			return;
+		}
+
+		if (app->size < sizeof(MultiMoveItem_Struct)) {
+			return; // Not enough data to be a valid packet
```
Comment (fryguy503) [inline zone/client_packet.cpp:10868]: if we return should we kick else there is a chance of desync and item loss? I assume the client is still honoring the attempt to move.

Comment (catapultam-habeo) [reply zone/client_packet.cpp:10868]: Yes, should probably go linkdead here instead.

File: zone/client_packet.cpp [hunk 2]
```diff
@@ -10859,7 +10859,50 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+		}
+
+		const MultiMoveItem_Struct* multi_move = reinterpret_cast<const MultiMoveItem_Struct*>(app->pBuffer);
+		if (app->size != sizeof(MultiMoveItem_Struct) + sizeof(MultiMoveItemSub_Struct) * multi_move->count) {
+			return; // Packet size does not match expected size
```
Comment (fryguy503) [inline zone/client_packet.cpp:10873]: same as above

File: zone/client_packet.cpp (hunk 1 above)
Comment (catapultam-habeo) [reply zone/client_packet.cpp:10868]: HandleOP_MoveItem handles these situations by just returning, which is a similar desync situation.

**Comment by catapultam-habeo** [comment]
Do not merge this. Found some bugs.

**Comment by catapultam-habeo** [comment]
This isn't working correctly because the cursor bag re-orders itself on SwapItem, causing the transfers subsequent to the first one to fail. I'm not certain how to fix that right now.

**Comment by catapultam-habeo** [comment]
This works according to my testing. It implements all 3 ctrl-click modes exactly replicating client behavior.

1) ctrl + left click single item into bag -> places item into bag without opening it. This will combine stacks as needed, including filling multiple stacks as needed.
2) ctrl + left click bag into bag -> fills target bag with contents of source bag. client checks if there is enough room, etc. This will combine stacks as possible\needed.
3) ctrl + right click bag into bag -> swaps contents of bags. This puts things into lowest ordinal bag slots, but does not combine stacks. 

left and right click logic needs to be bifurcated because left click handles things a lot like proper slot swaps or autoinv in terms of stack combining, etc, so we want to leverage the existing logic of SwapItem. right click just bluntly exchanges contents, but we need to respect the new targets instead of just explicitly exchanging the bag contents slot for slot.

File: zone/client_packet.cpp [hunk 3]
```diff
@@ -10859,7 +10859,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+				uint16 to_slot;
+			};
+
+			std::vector<MoveInfo> items;
+    		items.reserve(multi_move->count);
+
+			for (int i = 0; i < multi_move->count; i++) {
+				// These are always bags, so we don't need to worry about raw items in slotCursor
+				auto from_slot   = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot, multi_move->moves[i].from_slot.SubIndex);
+				if (multi_move->moves[i].from_slot.Type == 1) { // Target is bank inventory
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10939]: Let’s make constants for these types to avoid magic numbers.

File: zone/client_packet.cpp [hunk 4]
```diff
@@ -10859,7 +10859,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+					from_slot = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot + EQ::invslot::BANK_BEGIN, multi_move->moves[i].from_slot.SubIndex);
+				} else if (multi_move->moves[i].from_slot.Type == 2) { // Target is shared bank inventory
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10941]: Let’s make constants for these types to avoid magic numbers.

File: zone/client_packet.cpp [hunk 5]
```diff
@@ -10859,7 +10859,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+					from_slot = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot + EQ::invslot::SHARED_BANK_BEGIN, multi_move->moves[i].from_slot.SubIndex);
+				}
+
+				auto to_slot   = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot, multi_move->moves[i].to_slot.SubIndex);
+				if (multi_move->moves[i].to_slot.Type == 1) { // Target is bank inventory
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10946]: Let’s make constants for these types to avoid magic numbers.

File: zone/client_packet.cpp [hunk 6]
```diff
@@ -10859,7 +10859,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+					to_slot = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot + EQ::invslot::BANK_BEGIN, multi_move->moves[i].to_slot.SubIndex);
+				} else if (multi_move->moves[i].to_slot.Type == 2) { // Target is shared bank inventory
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10948]: Let’s make constants for these types to avoid magic numbers.

File: zone/client_packet.cpp [hunk 7]
```diff
@@ -10859,7 +10859,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+		}
+
+		// This is a left click which is purely additive. This should always be cursor object or cursor bag contents into general\bank\whatever bag
+		if (left_click) {
+			for (int i = 0; i < multi_move->count; i++) {
+				MoveItem_Struct* mi = new MoveItem_Struct();
+				mi->from_slot 	= multi_move->moves[i].from_slot.SubIndex == -1 ? multi_move->moves[i].from_slot.Slot : m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot, multi_move->moves[i].from_slot.SubIndex);									
+				mi->to_slot = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot, multi_move->moves[i].to_slot.SubIndex);
+				
+				if (multi_move->moves[i].to_slot.Type == 1) { // Target is bank inventory
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10904]: Let’s make constants for these types to avoid magic numbers.

File: zone/client_packet.cpp [hunk 8]
```diff
@@ -10859,7 +10859,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+					mi->to_slot = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot + EQ::invslot::BANK_BEGIN, multi_move->moves[i].to_slot.SubIndex);
+				} else if (multi_move->moves[i].to_slot.Type == 2) { // Target is shared bank inventory
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10906]: Let’s make constants for these types to avoid magic numbers.

File: zone/client_packet.cpp (hunk 3 above)
Comment (catapultam-habeo) [reply zone/client_packet.cpp:10939]: Where should these be defined?

Comment (Kinglykrab) [reply zone/client_packet.cpp:10939]: https://github.com/EQEmu/Server/blob/master/common/emu_constants.h

**Comment by catapultam-habeo** [comment]
This has been running on my servers for a while now without issues (~40 users total). I haven't been able to break it personally. I'm reasonably confident in this implementation.

File: common/emu_constants.h [hunk 9]
```diff
@@ -72,6 +72,12 @@ namespace EQ
 		int16 GetInvTypeSize(int16 inv_type);
 		using RoF2::invtype::GetInvTypeName;
 
+		enum MoveMultiType: int16 {
+			inventoryPersonal = 0,
+			inventoryBank = 1,
+			inventorySharedBank = 2
+		};
+
```
Comment (Kinglykrab) [inline common/emu_constants.h:80]: These are unused?

File: common/eq_packet_structs.h [hunk 10]
```diff
@@ -1619,6 +1619,32 @@ struct MoveItem_Struct
...
 
+// New for RoF2 - Size: 12
+struct InventorySlot_Struct
+{
+/*000*/	int16	Type;		// Worn and Normal inventory = 0, Bank = 1, Shared Bank = 2, Delete Item = -1
+/*002*/	int16	Unknown02;
+/*004*/	int16	Slot;
+/*006*/	int16	SubIndex;
+/*008*/	int16	AugIndex;	// Guessing - Seen 0xffff
+/*010*/	int16	Unknown01;	// Normally 0 - Seen 13262 when deleting an item, but didn't match item ID
```
Comment (Kinglykrab) [inline common/eq_packet_structs.h:1630]: `snake_case` please.

File: zone/client_packet.cpp [hunk 11]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
 
 void Client::Handle_OP_MoveMultipleItems(const EQApplicationPacket *app)
 {
-	Kick("Unimplemented move multiple items"); // TODO: lets not desync though
+	// This packet is only sent from the client if we ctrl click items in inventory
+	if (m_ClientVersionBit & EQ::versions::maskRoF2AndLater) {
+
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10883]: Unnecessary newline.

File: zone/client_packet.cpp [hunk 12]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+			for (int i = 0; i < multi_move->count; i++) {
+				// These are always bags, so we don't need to worry about raw items in slotCursor
+				auto from_slot   = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot, multi_move->moves[i].from_slot.SubIndex);
+				if (multi_move->moves[i].from_slot.Type == EQ::invtype::inventoryBank) { // Target is bank inventory
+					from_slot = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot + EQ::invslot::BANK_BEGIN, multi_move->moves[i].from_slot.SubIndex);
+				} else if (multi_move->moves[i].from_slot.Type == EQ::invtype::inventorySharedBank) { // Target is shared bank inventory
+					from_slot = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot + EQ::invslot::SHARED_BANK_BEGIN, multi_move->moves[i].from_slot.SubIndex);
+				}
+
+				auto to_slot   = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot, multi_move->moves[i].to_slot.SubIndex);
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10964]: Can use actual type here since it's a simple type.

File: zone/client_packet.cpp [hunk 13]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+				EQ::ItemInstance* item;
+				uint16 to_slot;
+			};
+
+			std::vector<MoveInfo> items;
+    		items.reserve(multi_move->count);
+
+			for (int i = 0; i < multi_move->count; i++) {
+				// These are always bags, so we don't need to worry about raw items in slotCursor
+				auto from_slot   = m_inv.CalcSlotId(multi_move->moves[i].from_slot.Slot, multi_move->moves[i].from_slot.SubIndex);
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10957]: Can use actual type here since it's a simple type.

File: zone/client_packet.cpp [hunk 14]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+		}
+
+		const MultiMoveItem_Struct* multi_move = reinterpret_cast<const MultiMoveItem_Struct*>(app->pBuffer);
+		if (app->size != sizeof(MultiMoveItem_Struct) + sizeof(MultiMoveItemSub_Struct) * multi_move->count) {
+			LinkDead();
+			return; // Packet size does not match expected size
+		}
+			
+		const auto from_parent = multi_move->moves[0].from_slot.Slot;
+		const auto to_parent   = multi_move->moves[0].to_slot.Slot;
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10901]: Can use actual types here since they're simple types.

File: zone/client_packet.cpp [hunk 15]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+			LinkDead();
+			return;
+		}
+
+		if (app->size < sizeof(MultiMoveItem_Struct)) {
+			LinkDead();
+			return; // Not enough data to be a valid packet
+		}
+
+		const MultiMoveItem_Struct* multi_move = reinterpret_cast<const MultiMoveItem_Struct*>(app->pBuffer);
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10894]: Could use `const auto` if you'd prefer since it's a complex type.

File: zone/client_packet.cpp [hunk 16]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+		for (int i = 0; i < multi_move->count; i++) {
+			if (multi_move->moves[i].from_slot.Slot != EQ::invslot::slotCursor) {
+				left_click = false;
+			}
+		}
+
+		// This is a left click which is purely additive. This should always be cursor object or cursor bag contents into general\bank\whatever bag
+		if (left_click) {
+			for (int i = 0; i < multi_move->count; i++) {
+				MoveItem_Struct* mi = new MoveItem_Struct();
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10919]: Could use `auto` if you'd prefer since it's a complex type.

File: zone/client_packet.cpp [hunk 17]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+				if (move.item) {				
+					items.push_back(move);
+				}
+			}
+
+			for (const MoveInfo& move : items) {
+				PutItemInInventory(move.to_slot, *move.item); // This saves inventory too
+			}
+		}
+
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10984]: Unnecessary newline.

File: zone/client_packet.cpp [hunk 18]
```diff
@@ -10878,7 +10878,114 @@ void Client::Handle_OP_MoveItem(const EQApplicationPacket *app)
...
+				if (multi_move->moves[i].to_slot.Type == EQ::invtype::inventoryBank) { // Target is bank inventory
+					to_slot = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot + EQ::invslot::BANK_BEGIN, multi_move->moves[i].to_slot.SubIndex);
+				} else if (multi_move->moves[i].to_slot.Type == EQ::invtype::inventorySharedBank) { // Target is shared bank inventory
+					to_slot = m_inv.CalcSlotId(multi_move->moves[i].to_slot.Slot + EQ::invslot::SHARED_BANK_BEGIN, multi_move->moves[i].to_slot.SubIndex);
+				}
+
+				// Probably need some error checking here, but I wasn't able to produce any error states on purpose.				
+				MoveInfo move;
+				move.item = m_inv.PopItem(from_slot); // Don't delete the instance here
+				move.to_slot = to_slot;
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:10974]: Can do this.
```cpp
MoveInfo move{
    .item = m_inv.PopItem(from_slot), // Don't delete the instance here
    .to_slot = to_slot
};

if (move.item) {
```

File: common/eq_packet_structs.h (hunk 10 above)
Comment (catapultam-habeo) [reply common/eq_packet_structs.h:1630]: This exists already in `rof_structs.h`, I just copied it. Would it be more correct to adjust this in `rof_structs.h` and reference it from there?

Comment (Kinglykrab) [reply common/eq_packet_structs.h:1630]: Sure, whatever works.

**Comment by Akkadius** [comment]
@catapultam-habeo can you attach some demo videos?

Do you have any concerns of item dupes or item loss?

**Comment by catapultam-habeo** [comment]
> @catapultam-habeo can you attach some demo videos?
> 
> Do you have any concerns of item dupes or item loss?

Sorry for the delay in having KK's comments addressesed.

I'll get some demo videos when I make those adjustments and have a new build.

As for concerns about item dupes or loss; I don't see how either could happen unless there is a flaw with the existing inventory manipulation methods. That said, I think that there *is* a flaw, but I haven't been able to locate it. Unrelated to this PR (I have observed this behavior prior to working on this, on a totally stock release server), items can dup into ostensibly unaddressable inventory Slot IDs. There doesn't seem to be any real consequence other than extraneous `inventory` table rows, as these items do appear to be inaccessible.

**Comment by catapultam-habeo** [comment]
That was a rebase.

File: common/eq_packet_structs.h (hunk 10 above)
Comment (catapultam-habeo) [reply common/eq_packet_structs.h:1631]: I do not want to own a change that touches as much code as would be affected by this.

**Comment by catapultam-habeo** [comment]
I'm comfortable with this. I... don't actually know how to make or post videos to demo this in action. It is currently deployed on Retribution.

**Comment by Akkadius** [comment]
> I'm comfortable with this. I... don't actually know how to make or post videos to demo this in action. It is currently deployed on Retribution.

You can use [OBS](https://obsproject.com/)

**Comment by catapultam-habeo** [comment]
[Uploading Screencast from 07-07-2024 10:01:57 PM.webm…]()

**Comment by catapultam-habeo** [comment]
[Screencast from 07-07-2024 10:01:57 PM.webm](https://github.com/EQEmu/Server/assets/97849758/9c21aa9b-14e7-471f-aa90-224462b1d18f)

**Comment by catapultam-habeo** [comment]
I found a dupe. Stand by, do not merge.

**Comment by catapultam-habeo** [comment]
I have corrected the duplication bug. Attached is a video test of all functions. 

[Screencast from 07-08-2024 05:17:06 PM.webm](https://github.com/EQEmu/Server/assets/97849758/9453c7dc-711c-4cd2-bc43-14b595571ee5)

File: common/eq_packet_structs.h (hunk 10 above)
Comment (Akkadius) [reply common/eq_packet_structs.h:1631]: Fair

**Comment by Akkadius** [comment]
> I have corrected the duplication bug. Attached is a video test of all functions.
> 
>  Screencast.from.07-08-2024.05.17.06.PM.webm

Thank you for addressing all of the changes and posting the video. I think this is great you put this feature together. 

My only concern (albeit large) was the feedback around item loss and dupes. 

If we feel like we are good there - then I'm good.

**Comment by catapultam-habeo** [comment]
The previous issue was caused by `PopItem(...)` not sending database updates, so bag swaps could cause additional items to be saved into db in certain circumstances.

I haven't been able to cause any more issues like this; asymmetrical bags, full bags, etc.

**Comment by fryguy503** [comment]
I pulled this to WFH for live testing, will post any issues I encounter if any.

**Comment by Akkadius** [comment]
> I pulled this to WFH for live testing, will post any issues I encounter if any.

Awesome

**Review by fryguy503 — APPROVED** [review]
Been running this for a week and have not noticed any issues related to this PR. Players that use it absolutely love it!

Thank you Cata!

**Pull request merged**